File: src/Web/TypeScriptTS/example2.ts
// Abstract Product
interface Database {
    connect(): void;
    query(): void;
}

// Concrete Product
class Postgres implements Database {
    connect(): void {
        console.log("Connecting to PostgreSQL");
    }

    query(): void {
        console.log("Querying PostgreSQL");
    }
}

class MySQL implements Database {
    connect(): void {
        console.log("Connecting to MySQL");
    }

    query(): void {
        console.log("Querying MySQL");
    }
}

// Abstract Factory
interface DatabaseFactory {
    createDatabase(): Database;
}

// Concrete Factory
class PostgresFactory implements DatabaseFactory {
    createDatabase(): Database {
        return new Postgres();
    }
}

class MySQLFactory implements DatabaseFactory {
    createDatabase(): Database {
        return new MySQL();
    }
}

function useDatabase(factory: DatabaseFactory): void {
    const db = factory.createDatabase();
    db.connect();
    db.query();
}

// Usage
useDatabase(new PostgresFactory());
useDatabase(new MySQLFactory());

File: src/DataScience/Julia/example2.jl
abstract type Database end

struct Postgres <: Database end
struct MySQL <: Database end

function connect(db::Postgres)
    println("Connecting to PostgreSQL")
end

function query(db::Postgres)
    println("Querying PostgreSQL")
end

function connect(db::MySQL)
    println("Connecting to MySQL")
end

function query(db::MySQL)
    println("Querying MySQL")
end

abstract type DatabaseFactory end

struct PostgresFactory <: DatabaseFactory end
struct MySQLFactory <: DatabaseFactory end

function create_database(::PostgresFactory)::Database
    return Postgres()
end

function create_database(::MySQLFactory)::Database
    return MySQL()
end

function use_database(factory::DatabaseFactory)
    db = create_database(factory)
    connect(db)
    query(db)
end

# Usage
use_database(PostgresFactory())
use_database(MySQLFactory())
